### PR TITLE
CASMCMS-7676 Add documentation for IMS recipe templating

### PR DESCRIPTION
**Pull request by ecozzi-hpe**
## Summary and Scope

_Summarize what has changed. Explain why this PR is necessary. What is impacted? Is this a new feature, critical bug fix, etc?_

Add docs for creating IMS recipes that get templated during IMS create jobs. 

_Is this change backwards incompatible, backwards compatible, or a backwards compatible bugfix?_
Yes

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Resolves [CASMCMS-7676](https://jira-pro.its.hpecorp.net:8443/browse/CASMCMS-7676)

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable
- [ ] Do not merge until CSM 1.2.5 content can be merged.

### CODE REVIEW

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 1]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
     type = "s3"
     ```
 
+## Templating IMS recipes when building an IMS image
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:134]: Change to:

Templating IMS Recipes when Building an IMS Image

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 2]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+
+IMS can optionally template the contents of an IMS recipe before building the recipe with kiwi-ng during an IMS create job. This
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:136]: kiwi-ng -> Kiwi-NG

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 3]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
     type = "s3"
     ```
 
+## Templating IMS recipes when building an IMS image
+
+IMS can optionally template the contents of an IMS recipe before building the recipe with kiwi-ng during an IMS create job. This 
+enables variables stored in the recipe to be dynamically replaced with values registered with the IMS recipe record after the 
+recipe is downloaded from S3 but before it is built by Kiwi-ng. This functionality is used by various Cray products to ensure
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:138]: - Add comma after S3
- Kiwi-ng -> Kiwi-NG

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 4]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+   -rw-r--r-- 1 root root    29 Feb  4 21:14 .ims_recipe_template.yaml
+   drwxr-xr-x 3 root root    18 Feb  4 21:15 root
+      
+   ncn# cat .ims_recipe_template.yaml
+   template_files:
+   - config.xml
+   ```
+   
+   IMS will template the indicated files 'in place' -- that is, the listed file will be replaced with the templated copy.
+   NOTE: Only files listed under the template_files key will be templated by IMS.
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:166]: NOTE:   ->   **NOTE:**

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 5]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+Follow the steps below to enable IMS templating within a given recipe. Note that for IMS to properly template a recipe,
+the following procedures must all be completed.
+
+### Enable templating in IMS recipe archive tgz file
+
+1. Add a file named '.ims_recipe_template.yaml' to the root of the recipe archive. The contents of this file must be valid
+   yaml containing the keyword `template_files`. The `template_files` keyword must contain a list of files to be templated
+   within the image recipe archive. 
+
+   ``` bash
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:150]: In general, input/output should be separated. For example:

```
ncn# ls -al
```

Example output:

```
   drwxr-xr-x 3 root root    70 Feb 22 19:05 .
   drwxr-xr-x 3 root root    41 Feb 22 17:18 ..      
   -rw-r--r-- 1 root root  2744 Feb  4 21:14 config.sh
   -rw-r--r-- 1 root root 24645 Feb  4 21:15 config.xml
   -rw-r--r-- 1 root root  1216 Feb  4 21:14 images.sh
   -rw-r--r-- 1 root root    29 Feb  4 21:14 .ims_recipe_template.yaml
   drwxr-xr-x 3 root root    18 Feb  4 21:15 root
```

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 6]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+   ```bash
+   ncn# cat config.xml
+   ...
+   <repository type="rpm-md" alias="csm-{{ CSM_RELEASE_VERSION }}-sle-15sp3" priority="2" imageinclude="true">
+       <source path="https://packages.local/repository/csm-{{ CSM_RELEASE_VERSION }}-sle-15sp3/"/>
+   </repository>
+   ...
+   ```
+
+   NOTE: The repo referenced above is for documentation purposes only and may not actually exist.
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:181]: NOTE: -> **NOTE:**

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 7]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+
+### Building an image from an IMS templated recipe
+
+The procedure to build an image from an IMS recipe that uses templating does not change. Follow the normal IMS create
+procedure, specifying the recipe's ID value in the job's `--artifact-id` parameter. The IMS job will start as normal,
+but after downloading the recipe from S3, you will see an indication that IMS is templating the recipe in the job's
+`fetch-recipe` container.
+
+```
+ncn-m001:~ # i logs cray-ims-812e9cda-62a8-4d1b-83a3-5891340671ff-create-78z6p -c fetch-recipe
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:260]: - command prompt should be ncn-m001#
- same comment as above about input/output

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 8]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+   key = "CSM_RELEASE_VERSION"
+   value = "1.2.5"
+   
+   [link]
+   path = "s3://ims/recipes/2233c82a-5081-4f67-bec4-4b59a60017a6/recipe.tar.gz"
+   etag = ""
+   type = "s3"
+   ```
+
+### Building an image from an IMS templated recipe
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:252]: Change to:

Build an Image from an IMS Templated Recipe

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 9]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+
+3. Create a tgz archive of the image recipe.
+
+    ```bash
+    ncn# export ARTIFACT_FILENAME=recipe.tar.gz
+    ncn# tar cvfz ../$ARTIFACT_FILE .
+    ncn# cd ..
+    ```
+
+### Add template key/value pairs to IMS recipe record
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:191]: Change to:

Add Template Key/Value Pairs to IMS Recipe Record

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 10]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+
+IMS can optionally template the contents of an IMS recipe before building the recipe with kiwi-ng during an IMS create job. This 
+enables variables stored in the recipe to be dynamically replaced with values registered with the IMS recipe record after the 
+recipe is downloaded from S3 but before it is built by Kiwi-ng. This functionality is used by various Cray products to ensure 
+that the resulting image is built from the most correct release versions of Nexus hosted RPM repositories. 
+
+Follow the steps below to enable IMS templating within a given recipe. Note that for IMS to properly template a recipe,
+the following procedures must all be completed.
+
+### Enable templating in IMS recipe archive tgz file
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:144]: Change to:

Enable Templating in IMS Recipe Archive tgz File

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 11]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+    name = "my_recipe.tgz"
+    recipe_type = "kiwi-ng"
+    
+    [[template_dictionary]]
+    key = "CSM_RELEASE_VERSION"
+    value = "1.2.5"
+    ```
+
+    Additional key/value pairs can be added by providing a list of comma-separated keys/values to the 
+    --template-dictionary-key and --template-dictionary-value parameters.
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:213]: enclose parameters in backticks

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 12]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+    ncn# export ARTIFACT_FILENAME=recipe.tar.gz
+    ncn# tar cvfz ../$ARTIFACT_FILE .
+    ncn# cd ..
+    ```
+
+### Add template key/value pairs to IMS recipe record
+
+1. Create a new IMS recipe record with template_dictionary key/value pairs.
+
+    ```bash
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:195]: same comment as above about input/output

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 13]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+
+   It is suggested as a best practice that the S3 object name start with `recipes/` and contain the IMS recipe ID to remove ambiguity.
+
+   ```bash
+   ncn# cray artifacts create ims recipes/$IMS_RECIPE_ID/$ARTIFACT_FILENAME $ARTIFACT_FILENAME
+   ```
+
+3. Update the IMS recipe record with the S3 path to the recipe archive.
+
+   ```bash
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:231]: same comment as above about input/output

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 14]
```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+   path = "s3://ims/recipes/2233c82a-5081-4f67-bec4-4b59a60017a6/recipe.tar.gz"
+   etag = ""
+   type = "s3"
+   ```
+
+### Building an image from an IMS templated recipe
+
+The procedure to build an image from an IMS recipe that uses templating does not change. Follow the normal IMS create
+procedure, specifying the recipe's ID value in the job's `--artifact-id` parameter. The IMS job will start as normal,
+but after downloading the recipe from S3, you will see an indication that IMS is templating the recipe in the job's
```
Comment (nrockershousen) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:256]: Change "you will see an" to "there will be an"

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 1 above)
Comment (ecozzi-hpe) [reply operations/image_management/Upload_and_Register_an_Image_Recipe.md:134]: Done

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 2 above)
Comment (ecozzi-hpe) [reply operations/image_management/Upload_and_Register_an_Image_Recipe.md:136]: Done

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 3 above)
Comment (ecozzi-hpe) [reply operations/image_management/Upload_and_Register_an_Image_Recipe.md:138]: Done

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 5 above)
Comment (ecozzi-hpe) [reply operations/image_management/Upload_and_Register_an_Image_Recipe.md:150]: Done

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 4 above)
Comment (ecozzi-hpe) [reply operations/image_management/Upload_and_Register_an_Image_Recipe.md:166]: Done

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 10 above)
Comment (ecozzi-hpe) [reply operations/image_management/Upload_and_Register_an_Image_Recipe.md:144]: Done

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 11 above)
Comment (ecozzi-hpe) [reply operations/image_management/Upload_and_Register_an_Image_Recipe.md:213]: Done

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 13 above)
Comment (ecozzi-hpe) [reply operations/image_management/Upload_and_Register_an_Image_Recipe.md:231]: done

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 14 above)
Comment (ecozzi-hpe) [reply operations/image_management/Upload_and_Register_an_Image_Recipe.md:256]: done

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 6 above)
Comment (ecozzi-hpe) [reply operations/image_management/Upload_and_Register_an_Image_Recipe.md:181]: Done

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 7 above)
Comment (ecozzi-hpe) [reply operations/image_management/Upload_and_Register_an_Image_Recipe.md:260]: Done

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 8 above)
Comment (ecozzi-hpe) [reply operations/image_management/Upload_and_Register_an_Image_Recipe.md:252]: Done

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 9 above)
Comment (ecozzi-hpe) [reply operations/image_management/Upload_and_Register_an_Image_Recipe.md:191]: Done

**Review by nrockershousen — APPROVED** [review]

**Comment by ecozzi-hpe** [comment]
Need to rebase code for release/1.2 Will open new PR.

**Pull request closed without merging**